lc0/src/neural/network_cudnn.cu

```diff
@@ -0,0 +1,836 @@
+            network.push_back(conv1);
+
+            ConvLayer *conv2 = new ConvLayer(getLastLayer(), numFilters, 8, 8, 3, numFilters, true, true);
+            conv2->loadWeights(&weights.residual[block].conv2.weights[0], &weights.residual[block].conv2.biases[0]);
+            network.push_back(conv2);
+        }
+
+        resiLast = getLastLayer();
+
+        // policy head
+        {
+            ConvLayer *convPol = new ConvLayer(resiLast, weights.policy.bn_means.size(), 8, 8, 1, numFilters);
+            convPol->loadWeights(&weights.policy.weights[0]);
+            network.push_back(convPol);
+
+            BNLayer *BNPol = new BNLayer(getLastLayer(), true);
+            BNPol->loadWeights(&weights.policy.bn_means[0], &weights.policy.bn_stddivs[0]);
+            network.push_back(BNPol);
+
+            FCLayer *FCPol = new FCLayer(getLastLayer(), weights.ip_pol_b.size(), 1, 1, false, true);
+            FCPol->loadWeights(&weights.ip_pol_w[0], &weights.ip_pol_b[0]);
+            network.push_back(FCPol);
+
+            SoftMaxLayer *softmaxPol = new SoftMaxLayer(getLastLayer());
+            network.push_back(softmaxPol);
+        }
+        policyOut = getLastLayer();
+
+        // Value head
+        {
+            ConvLayer *convVal = new ConvLayer(resiLast, weights.value.bn_means.size(), 8, 8, 1, numFilters);
+            convVal->loadWeights(&weights.value.weights[0]);
+            network.push_back(convVal);
+
+            BNLayer *BNVal = new BNLayer(getLastLayer(), true);
+            BNVal->loadWeights(&weights.value.bn_means[0], &weights.value.bn_stddivs[0]);
+            network.push_back(BNVal);
+
+            FCLayer *FCVal1 = new FCLayer(getLastLayer(), weights.ip1_val_b.size(), 1, 1, true, true);
+            FCVal1->loadWeights(&weights.ip1_val_w[0], &weights.ip1_val_b[0]);
+            network.push_back(FCVal1);
+
+            FCLayer *FCVal2 = new FCLayer(getLastLayer(), 1, 1, 1, false, true, true);
+            FCVal2->loadWeights(&weights.ip2_val_w[0], &weights.ip2_val_b[0]);
+            network.push_back(FCVal2);
+        }
+        valueOut = getLastLayer();
+
+        // 2. allocate GPU memory for running the network
+        //    - three buffers of max size are enough (one to hold input, second to hold output and third to hold skip connection's input)
+        resiLast->setBatchSize(MAX_BATCH_SIZE);
+        size_t maxSize = resiLast->getOutputSize();
+        for (int i = 0; i < 3; i++)
+        {
+            cudaMalloc(&tensorMem[i], maxSize);
+            cudaMemset(tensorMem[i], 0, maxSize);
+        }
+
+        //printf("Allocated %d bytes of GPU memory to run the network\n", 3 * maxSize);
+
+        // 3. allocate scratch space (used internally by cudnn to run convolutions)
+        cudaMalloc(&scratchMem, CUDA_SCRATCH_SIZE);
+
+    }
+
+    void forwardEval(float *input, float *op_pol, float *op_val, int batchSize) const
+    {
+        // printf(" ..%d.. ", batchSize);
+
+        g_lock.lock();
```

Comment:
Consider using std::lock_guard instead of locking and unlocking manually.

Reply:
done